### PR TITLE
Matching: go further on simpler types

### PR DESCRIPTION
**Pull request by gasche**
Note: I think you got rid of the `.ocamlformat` file along the way, so I cannot reformat on my side.

### CODE REVIEW

**Comment by trefis** [comment]
> Note: I think you got rid of the .ocamlformat file along the way, so I cannot reformat on my side.

Yeah, it got moved to lambda/ during a rebase on trunk, and then didn't get moved back when rerebasing on 4.08.
But no worries, I'll redo a `git filter-branch` before sending PRs upstream.

I will look at your commits tomorrow, but one thing I noticed is that you didn't start from my "rematch-pattern-head" branch that I had linked in my email (well, actually I had link [the diff](https://github.com/trefis/ocaml/compare/matching-3rd-pr-408...trefis:rematch-pattern-head) of that branch against the one you're starting off)
But no matter.

**Comment by trefis** [comment]
And I actually looked at the code, and realized you probably didn't start of my branch on purpose as there is no Pattern_head appearance yet.

Anyway, I like the changes/simplifications!

Do you want to keep pilling things on in this PR, or should I include it in my branch and you'll send other PRs ?

**Comment by gasche** [comment]
I'll include these commits in another PR with more stuff (including a rebased version of your change on top of this head).

**Comment by trefis** [comment]
Ack. Shall I just merge this then?

**Comment by gasche** [comment]
Yes, I think that would be nice -- if you like the code overall.

**Comment by trefis** [comment]
I do. Done.

**Comment by gasche** [comment]
Note: I think I found a small bug in my most recent head (running make core twice has the second attempt failing on int32.ml with what looks like a miscompilation bug), but I'm not sure when (and by whom) it was introduced. This will have to be debugged at some point.

**Comment by trefis** [comment]
Hum, that's surprising. I used your script (which I believes makes core twice with one after a bootstrap) earlier to produce all the lambda dumps for the various branches.
So I guess it's in these patches? I will have a more serious look tomorrow then.

File: bytecomp/matching.ml [hunk 1]
```diff
@@ -1125,25 +1127,22 @@ let as_matrix pat_of_head cases =
 
 *)
 
-let rec split_or argo (cls : Half_simplified_clause.t list) args def =
-  let rec do_split rev_before rev_ors rev_no = function
+let rec split_or argo (cls : Half_simple.clause list) args def =
+  let rec do_split (rev_before : Simple.clause list) rev_ors rev_no = function
     | [] ->
-        cons_next
-          (List.rev_map Simplified_clause.assert_no_or rev_before)
```
Comment (trefis) [inline bytecomp/matching.ml:1132]: That `rev_map` simply disappears, it should have stayed as a simple `rev`.

**Pull request merged**